_posts/2021-04-09-springboot-autoconfigurations.md

```diff
@@ -0,0 +1,436 @@
+
+The below code snippet shows the configuration of Dispatcher Servlet in a Web Application in ``web.xml``
+
+```xml
+
+    <servlet>
+        <servlet-name>dispatcher</servlet-name>
+        <servlet-class>
+            org.springframework.web.servlet.DispatcherServlet
+        </servlet-class>
+        <init-param>
+            <param-name>contextConfigLocation</param-name>
+            <param-value>/WEB-INF/musicservice-servlet.xml</param-value>
+        </init-param>
+        <load-on-startup>1</load-on-startup>
+    </servlet>
+
+    <servlet-mapping>
+        <servlet-name>dispatcher</servlet-name>
+        <url-pattern>/</url-pattern>
+    </servlet-mapping>
+
+```
+For delivering static content we would InternalViewResolver bean needs to be configured
+
+```xml
+
+   <bean
+        class="org.springframework.web.servlet.view.InternalResourceViewResolver">
+        <property name="prefix">
+            <value>/WEB-INF/views/</value>
+        </property>
+        <property name="suffix">
+            <value>.jsp</value>
+        </property>
+  </bean>
+
+  <mvc:resources mapping="/webjars/**" location="/webjars/"/>
+
+```
+When we use Hibernate/JPA, we would need to configure a data source, an entity manager factory, a transaction manager factory manually.
+
+```xml
+
+    <bean id="dataSource" class="com.mchange.v2.c3p0.ComboPooledDataSource"
+        destroy-method="close">
+        <property name="driverClass" value="${db.driver}" />
+        <property name="jdbcUrl" value="${db.url}" />
+        <property name="user" value="${db.username}" />
+        <property name="password" value="${db.password}" />
+    </bean>
+
+    <jdbc:initialize-database data-source="dataSource">
+        <jdbc:script location="classpath:config/schema.sql" />
+        <jdbc:script location="classpath:config/data.sql" />
+    </jdbc:initialize-database>
+
+    <bean
+        class="org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean"
+        id="entityManagerFactory">
+        <property name="persistenceUnitName" value="musicservice_pu" />
+        <property name="dataSource" ref="dataSource" />
+    </bean>
+
+    <bean id="transactionManager" class="org.springframework.orm.jpa.JpaTransactionManager">
+        <property name="entityManagerFactory" ref="entityManagerFactory" />
+        <property name="dataSource" ref="dataSource" />
+    </bean>
+
+    <tx:annotation-driven transaction-manager="transactionManager"/>
+
+ ```
+In Spring-based application, above XML configuration we have to do manually for each of the beans contained in it and wire/associate them properly.
+
+## 4. Role of Conditional Auto-Configurations with @Conditional annotation
+
+Before we understand the concept of auto-configuration, we need to understand just one important concept i.e., Spring Framework's ``@Conditional`` annotation.
+
+### 4.1 What is @Conditional?
```

Comment:
A section with only a single sub-section is usually a sign that you can remove the sub section